### PR TITLE
Sxfm constraints

### PR DESCRIPTION
**Pull request by ChristianKaltenecker**
This is related to https://github.com/se-sic/VaRA/issues/773

Furthermore, I have fixed an issue in the syntax diagram of the constraints

### CODE REVIEW

**Comment by ChristianKaltenecker** [comment]
The failed test depends on PR https://github.com/se-sic/vara-feature/pull/61.

File: include/vara/Feature/FeatureModelParser.h [hunk 1]
```diff
@@ -225,6 +225,8 @@ class FeatureModelSxfmParser : public FeatureModelParser {
   std::string Sxfm;
   FeatureModelBuilder FMB;
   std::string Indentation = "\t";
+  std::map<std::string, std::string> IdentifierMap =
+      std::map<std::string, std::string>();
```
Comment (vulder) [inline include/vara/Feature/FeatureModelParser.h]:
```suggestion
  std::map<std::string, std::string> IdentifierMap;
```

File: lib/Feature/FeatureModelParser.cpp [hunk 2]
```diff
@@ -540,6 +539,17 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
       }
 
+      // Search for identifier
+      std::regex Regex(R"(\([a-zA-Z\d_]*\))");
+      std::string CurrentLine = ToStringRef.str();
+      for (std::sregex_iterator Match = std::sregex_iterator(
+               CurrentLine.begin(), CurrentLine.end(), Regex);
+           Match != std::sregex_iterator(); Match++) {
+        std::string Identifier = (*Match).str();
+        Identifier = Identifier.substr(1, Identifier.size() - 2);
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp:563]: I get why you start at char 1 (to skip `:`) but why do you cut of 2 afterwards (1 for size() - 1, but where does the other character come from?)

File: lib/Feature/FeatureModelParser.cpp [hunk 3]
```diff
@@ -540,6 +539,17 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+        IdentifierMap[Identifier] = Name;
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp:564]:
```suggestion
        IdentifierMap[Identifier] = std::move(Name);
```

File: lib/Feature/FeatureModelParser.cpp [hunk 4]
```diff
@@ -609,8 +619,64 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+        << "Failed to read in constraints. Is the constraint field empty?\n";
+    return false;
+  }
+
+  std::stringstream Ss(reinterpret_cast<const char *>(
+      UniqueXmlChar(xmlNodeGetContent(Constraints), xmlFree).get()));
+  std::string To;
+
+  // Prepare the identifiers for later replacement
+  std::vector<std::string> Keys{};
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion
  std::vector<std::string> Keys;
  Keys.reserve(IdentifierMap.size());
```

File: lib/Feature/FeatureModelParser.cpp [hunk 5]
```diff
@@ -609,8 +619,64 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+  for (const auto &[key, value] : IdentifierMap) {
+    auto It = Keys.begin();
+    Keys.insert(It, key);
+  }
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion
  std::for_each(IdentifierMap.begin(), IdentifierMap.end(), [&Keys](const auto &Entry) { Keys.push_back(Entry.first); });
```

File: lib/Feature/FeatureModelParser.cpp [hunk 6]
```diff
@@ -609,8 +619,64 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+  std::sort(Keys.begin(), Keys.end(),
+            [](const std::string &First, const std::string &Second) {
+              return First.size() > Second.size();
+            });
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion
            std::greater<std::string>{});
```

File: lib/Feature/FeatureModelParser.cpp [hunk 7]
```diff
@@ -609,8 +619,64 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+    }
+
+    // We ignore the indentation in the beginning of each line since the format
+    // does not force it
+    llvm::StringRef ToStringRef(To);
+
+    if (!ToStringRef.contains(':')) {
+      llvm::errs() << "Failed to read in a constraint since it does not "
+                      "contain a colon.";
+      llvm::errs() << "This violates the format.";
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion
      llvm::errs() << "This violates the SXFM format.";
```

File: lib/Feature/FeatureModelParser.cpp [hunk 8]
```diff
@@ -609,8 +619,64 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+
+    // In the following lines, we replace all identifiers by the real feature
+    // name
+    for (auto &Key : Keys) {
+      std::string Value = IdentifierMap[Key];
+      while (CnfFormula.find(Key) != std::string::npos) {
+        auto FoundPos = CnfFormula.find(Key);
+        CnfFormula =
+            CnfFormula.replace(FoundPos, FoundPos + Key.length(), Value);
+      }
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion
      auto FoundPos = CnfFormula.find(Key);
      while (FoundPos != std::string::npos) {
        CnfFormula =
            CnfFormula.replace(FoundPos, FoundPos + Key.length(), Value);
        FoundPos = CnfFormula.find(Key);
      }
```

File: unittests/Feature/FeatureModelSxfm.cpp [hunk 9]
```diff
@@ -34,6 +34,16 @@ TEST(SxfmParser, validation) {
             Feature::FeatureKind::FK_BINARY);
 }
 
+/// Here, we check whether a wrongly formatted sxfm file would be parsed
+TEST(FeatureModelParser, detectWornglyFormattedConstraint) {
+  auto FS = llvm::MemoryBuffer::getFileAsStream(
+      getTestResource("test_wrong_constraint_format.sxfm"));
+  EXPECT_TRUE(FS && "Input file could not be read.");
+  auto FM =
+      FeatureModelSxfmParser(FS.get()->getBuffer().str()).buildFeatureModel();
```
Comment (vulder) [inline unittests/Feature/FeatureModelSxfm.cpp]:
```suggestion

  auto FM =
      FeatureModelSxfmParser(FS.get()->getBuffer().str()).buildFeatureModel();

```

**Review by vulder — CHANGES_REQUESTED** [review]

File: lib/Feature/FeatureModelParser.cpp (hunk 2 above)
Comment (ChristianKaltenecker) [reply lib/Feature/FeatureModelParser.cpp:563]: The identifier is a string such as `(root)`. Therefore, I have to remove the beginning and the ending paranthesis. This is why I need `Identifier.size() - 2`.

File: lib/Feature/FeatureModelParser.cpp [hunk 10]
```diff
@@ -553,6 +553,21 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlNode *FeatureTree) {
...
+      for (std::sregex_iterator Match = std::sregex_iterator(
+               CurrentLine.begin(), CurrentLine.end(), Regex);
+           Match != std::sregex_iterator(); Match++) {
+        std::string Identifier = (*Match).str();
+        Identifier = Identifier.substr(1, Identifier.size() - 2);
+        IdentifierMap[Identifier] = Name;
+        if (Identifier != Name && Name.find(Identifier) != std::string::npos) {
+          llvm::errs() << "Name must not contain ID:'" << Name << "'\n";
+          return false;
+        }
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp:568]: Why do you assign this before checking?
If we would do the check before, we could move the name string into the map.

**Review by vulder — APPROVED** [review]
However, somehow the coverage tracking does not work. Let's check this after merging. From the website I see only that no base report exists.

File: lib/Feature/FeatureModelParser.cpp (hunk 10 above)
Comment (ChristianKaltenecker) [reply lib/Feature/FeatureModelParser.cpp:568]: The issue is that I am using the variable `Name` later for creating a feature with this name using the `FeatureModelBuilder`, so `std::move` is no option here. 
But, of course, I could move the check a line earlier.

Comment (vulder) [reply lib/Feature/FeatureModelParser.cpp:568]: kk

**Pull request merged**